File: pset2/vigenere.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int main(int argc, string argv[])
{
    // ensure key is one word
    if (argc != 2)
    {
        printf("Usage: ./vigenere k\n");
        return 1;
    }
    
    // ensure key is alphabetical 
    for (int i = 0, n = strlen(argv[1]); i < n; i++)
    {
        if (!isalpha(argv[1][i]))
        {
            printf("Usage: ./vigenere k\n");
            return 1;
        }
    }
    
    // create alphabetical index for key
    
    int key[strlen(argv[1])];
    
    for (int i = 0, n = strlen(argv[1]); i < n; i++)
    {
        if (isupper(argv[1][i]))
        {
            key[i] = argv[1][i] - 65;
        }
        
        if (islower(argv[1][i]))
        {
            key[i] = argv[1][i] - 97;
        }
    }
    
    // get plaintext from user
    printf("plaintext: ");
    string plaintext = get_string();
    
    printf("ciphertext: ");
    
    // create countoff variable, alphabetical index and ciphertext array
    int j = 0;
    int alpha[strlen(plaintext)];
    int ciphertext[strlen(plaintext)];
    
    for (int i = 0, n = strlen(plaintext); i < n; i++)
    {
        // if p[i] is uppercase, convert to alpha index (-65) then add j'th element of key 
        if (isupper(plaintext[i]))
        {
            alpha[i] = plaintext[i] - 65;
            ciphertext[i] = (alpha[i] + key[j]) % 26;
            ciphertext [i] += 65;
            j++;
            printf("%c", ciphertext[i]);
        }
        
        // if p[i] is lowercase, convert to alpha index (-97) then add j'th element of key before 
        if (islower(plaintext[i]))
        {
            alpha[i] = plaintext[i] - 97;
            ciphertext[i] = (alpha[i] + key[j]) % 26;
            ciphertext [i] += 97;
            j++;
            printf("%c", ciphertext[i]);
        }
        
        // if p[i] is not alphabetical, print as normal
        if (!isalpha(plaintext[i]))
        {
            printf("%c", plaintext[i]);
        }
        
        // if final letter of key is hit, wrap around to the first letter
        if (j % strlen(argv[1]) == 0)
        {
            j = 0;
        }
    }
    
    //print new line
    printf("\n");
    return 0;
}

File: pset6/vigenere.py
# Final program used to ease into Python language.
# This was a real bugger to debug, but I learned 
# lots about how Python treats strings and lists,
# which are similar in structure to arrays, but 
# far, far more dynamic.

import cs50
import sys

# validate key
if len(sys.argv) != 2:
    print("usage: python vigenere.py k")
    exit(1)

# ensure key is alphabetical
if str.isalpha(sys.argv[1]) == False:
    print("usage: python vigenere.py k")
    exit(2)

# get plaintext from user
print("plaintext: ", end="")
plaintext = cs50.get_string()
print("ciphertext: ", end="")

# create alphabetical index for key
key = []
sys.argv[1] = str.lower(sys.argv[1])

for i in range(0, len(sys.argv[1])):
    key.append(ord(sys.argv[1][i]) - 97)

# create alphabetical index for plaintext, key
# tracking variable 'j' and ciphertext string
j = 0
ciphertext = []
alphatext = []

# encode plaintext and print new ciphertext
for i in range(0, len(plaintext)):
    
    if j % len(sys.argv[1]) == 0:
        j = 0
    
    if str.isupper(plaintext[i]) == True:
        alphatext.append(ord(plaintext[i]) - 65)
        ciphertext.append(chr(((alphatext[i] + key[j]) % 26) + 65))
        j += 1
        print("{}".format(ciphertext[i]), end="")
    
    if str.islower(plaintext[i]) == True:
        alphatext.append(ord(plaintext[i]) - 97)
        ciphertext.append(chr(((alphatext[i] + key[j]) % 26) + 97))
        j += 1
        print("{}".format(ciphertext[i]), end="")
    
    if str.isalpha(plaintext[i]) != True:
        alphatext.append(plaintext[i])
        ciphertext.append(plaintext[i])
        print("{}".format(ciphertext[i]), end="")

print("")
